teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,565 @@
+          digest:
+            type: string
+          digest_ago:
+            type: string
+    ```
+    And its corresponding TaskRun example will be like:
+    ``` yaml
+    TaskRun:
+      ...
+      results:
+        - name: SIGNABLE-VCS_FOO
+          value:
+            src_type: git
+            uri: git+https://github.com/foo/bar.git
+            digest_ago: sha256
+            digest: abc
+        - name: SIGNABLE-OCI_IMAGE_FOO
+          value:
+            name: gcr.io/somerepo/someimage
+            digest: abc
+            digest_ago: sha256
+    ```
+
+2. Use Perforce and images as sources, and generates in-toto provenance for a TaskRun that builds a Python wheel.
+
+    TODO: Example to be added 
+
+### Requirements
+
+- Able to structurally retrieve provenance info from TaskRun and produces currently supported provenance formats once a TaskRun is finished
+- Based on SLSA L2 requirements, the provenance must provide the location of source code in version control and the provenance MUST identify the output artifact via at least one cryptographic hash.
+- If a TaskRun is within a PipelineRun, the TaskRun’s provenance can be produced between the time that when one TaskRun is finished
+
+
+## Proposal
+
+Currently we are using type-hinting to retrieve needed provenance info from TaskRun’s results. The proposal aims to leverage the new features of structured results described in [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support better structured provenance retrieval for inputs and artifacts to make the current type-hinting easier. This proposal is focused on the use cases for generation in-toto attestations, as other types of attestations don’t need the TaskRun results.
+
+The workflow for the targeted results in Chains is like the following:
+```
+(TaskRun results) —-> (Scanned by Chains controller, targets are collected) —> (Targets are parsed and validated and matched against with defined result and param structures)  —-> (Chains controller sign the signable targets and put into intoto attestations)
+```
+Tekton Chains will follow [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support the sets of json schemas and described in the documents. In this proposal, the expected formats of results are also defined.
+
+As mentioned above, the signable targets can be split into inputs (sources) and outputs (artifacts). It's important that Chains are able to determine which signable targets are inputs, and which are outputs. We can enforce all the schemas need to have a string field `prov_type`, which can be either `input` or `output`, which indicates Chains to either put them into `subjects` or `materials`.
```

Comment:
I'm having trouble following how we're distinguishing inputs and outputs at the moment.

e.g. how would we distinguish "I used this image" vs "I pushed this image"?

Reply:
I'm trying to say that the result should have `prov_type` to indicate this. Are you trying to address another thing?

Reply:
Discussed offline. We are going to have SIGNABLE_INPUTS and SIGNABLE_OUTPUTs in results